File: results/dnd-rest-benchmark/lifecycle-runs/20260731T051216Z_claude_claude-sonnet-5_php-stdlib/lib/play_recipes.php
<?php
declare(strict_types=1);

// Routes: Play (campaign-scoped crafting recipes with deterministic
// ingredient requirements backed by the public inventory item catalog)
// ---------------------------------------------------------------------------

function validate_recipe_payload($body): ?array {
    if ($body === null
        || !isset($body['recipe_id'], $body['name'], $body['ingredients'], $body['output_item'], $body['output_quantity'])
        || !is_string($body['recipe_id']) || $body['recipe_id'] === ''
        || !is_string($body['name']) || $body['name'] === ''
        || !is_array($body['ingredients'])
        || !is_string($body['output_item']) || !in_array($body['output_item'], VALID_INVENTORY_ITEM_IDS, true)
        || !is_valid_int_range($body['output_quantity'], 1, PHP_INT_MAX)) {
        return null;
    }

    if (count($body['ingredients']) === 0) {
        return null;
    }

    $ingredients = [];
    foreach ($body['ingredients'] as $itemId => $quantity) {
        if (!is_string($itemId) || !in_array($itemId, VALID_INVENTORY_ITEM_IDS, true)) {
            return null;
        }
        if (!is_valid_int_range($quantity, 1, PHP_INT_MAX)) {
            return null;
        }
        $ingredients[$itemId] = (int)$quantity;
    }

    return [
        'recipe_id' => $body['recipe_id'],
        'name' => $body['name'],
        'ingredients' => $ingredients,
        'output_item' => $body['output_item'],
        'output_quantity' => (int)$body['output_quantity'],
    ];
}

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/recipes$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    if ($actor['username'] !== $campaign['owner']) {
        forbidden('only the campaign dm may create recipes');
    }

    $body = read_json_body();
    $validated = validate_recipe_payload($body);
    if ($validated === null) {
        bad_request();
    }
    $recipeId = $validated['recipe_id'];

    $stmt = $db->prepare('SELECT 1 FROM play_campaign_recipes WHERE campaign_id = ? AND recipe_id = ?');
    $stmt->execute([$campaignId, $recipeId]);
    if ($stmt->fetch(PDO::FETCH_ASSOC) !== false) {
        conflict('recipe id already exists in this campaign');
    }

    $recipe = [
        'recipe_id' => $recipeId,
        'name' => $validated['name'],
        'ingredients' => $validated['ingredients'],
        'output_item' => $validated['output_item'],
        'output_quantity' => $validated['output_quantity'],
    ];

    $stmt = $db->prepare('SELECT COALESCE(MAX(seq), 0) + 1 FROM play_campaign_recipes WHERE campaign_id = ?');
    $stmt->execute([$campaignId]);
    $seq = (int)$stmt->fetchColumn();

    $stmt = $db->prepare('INSERT INTO play_campaign_recipes (campaign_id, recipe_id, seq, data) VALUES (?, ?, ?, ?)');
    $stmt->execute([$campaignId, $recipeId, $seq, json_encode($recipe)]);

    send_json($recipe, 201);
}

if ($method === 'GET' && preg_match('#^/v1/play/campaigns/([^/]+)/recipes$#', $path, $pm)) {
    $campaignId = $pm[1];
    require_actor($db);

    require_play_campaign($db, $campaignId);

    $stmt = $db->prepare('SELECT data FROM play_campaign_recipes WHERE campaign_id = ? ORDER BY seq ASC');
    $stmt->execute([$campaignId]);
    $recipes = [];
    foreach ($stmt->fetchAll(PDO::FETCH_ASSOC) as $row) {
        $recipes[] = json_decode($row['data'], true);
    }

    send_json(['recipes' => $recipes], 200);
}

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/recipes/([^/]+)/craft$#', $path, $pm)) {
    $campaignId = $pm[1];
    $recipeId = $pm[2];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);
    if ($actor['username'] === $campaign['owner']) {
        forbidden('only a player may craft');
    }

    $stmt = $db->prepare('SELECT data FROM play_campaign_recipes WHERE campaign_id = ? AND recipe_id = ?');
    $stmt->execute([$campaignId, $recipeId]);
    $row = $stmt->fetch(PDO::FETCH_ASSOC);
    if ($row === false) {
        not_found('recipe not found');
    }
    $recipe = json_decode($row['data'], true);

    $body = read_json_body();
    if ($body === null || !isset($body['character_id']) || !is_string($body['character_id']) || $body['character_id'] === '') {
        bad_request();
    }
    $charId = $body['character_id'];

    $found = find_play_character($db, $campaignId, $charId);
    if ($found === null) {
        not_found('character not found');
    }
    [$username, $member] = $found;

    $owner = play_character_owner($db, $campaignId, $charId, $username);
    if ($actor['username'] !== $owner) {
        forbidden('only the character owner may craft for this character');
    }

    $items = $member['inventory_items'] ?? [];
    foreach ($recipe['ingredients'] as $itemId => $quantity) {
        if ((int)($items[$itemId] ?? 0) < $quantity) {
            conflict('insufficient ingredients');
        }
    }

    foreach ($recipe['ingredients'] as $itemId => $quantity) {
        $remaining = (int)($items[$itemId] ?? 0) - $quantity;
        if ($remaining > 0) {
            $items[$itemId] = $remaining;
        } else {
            unset($items[$itemId]);
        }
    }

    $outputItem = $recipe['output_item'];
    $items[$outputItem] = (int)($items[$outputItem] ?? 0) + $recipe['output_quantity'];
    $member['inventory_items'] = $items;

    $stmt = $db->prepare('UPDATE play_campaign_members SET data = ? WHERE campaign_id = ? AND username = ?');
    $stmt->execute([json_encode($member), $campaignId, $username]);

    send_json([
        'character_id' => $charId,
        'recipe_id' => $recipeId,
        'output_item' => $recipe['output_item'],
        'output_quantity' => $recipe['output_quantity'],
    ], 201);
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075550Z_pi_kimi-k2p7-code_ruby-sinatra/lib/routes/play_recipes.rb
# frozen_string_literal: true

# Route definitions for campaign crafting recipes.

# Returns the public recipe shape.
def recipe_response(recipe)
  {
    recipe_id: recipe[:recipe_id],
    name: recipe[:name],
    ingredients: recipe[:ingredients],
    output_item: recipe[:output_item],
    output_quantity: recipe[:output_quantity]
  }
end

# --- Create recipe ---
post '/v1/play/campaigns/:id/recipes' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  username = require_dm_actor!
  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_owner!(campaign, username)

  body = parse_json_body
  validate_recipe_body!(body)

  json_error(409, 'recipe already exists') if Storage.recipe_exists?(campaign_id, body['recipe_id'])

  Storage.create_recipe(
    campaign_id,
    body['recipe_id'],
    body['name'],
    body['ingredients'],
    body['output_item'],
    body['output_quantity']
  )

  status 201
  JSON.dump(recipe_response(Storage.load_recipe(campaign_id, body['recipe_id'])))
end

# --- List recipes ---
get '/v1/play/campaigns/:id/recipes' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  user = authenticate_actor!
  username = user[:username]

  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_access!(campaign, username)

  recipes = Storage.load_recipes(campaign_id).map { |r| recipe_response(r) }
  JSON.dump(recipes: recipes)
end

# --- Craft recipe ---
post '/v1/play/campaigns/:id/recipes/:recipe_id/craft' do
  content_type :json
  campaign_id = params[:id]
  recipe_id = params[:recipe_id]
  validate_campaign_id!(campaign_id)

  username = require_player_actor!

  load_play_campaign!(campaign_id)
  json_error(403, 'forbidden') unless Storage.play_campaign_member_exists?(campaign_id, username)

  body = parse_json_body
  validate_craft_body!(body)

  character_id = body['character_id']
  character = Storage.load_play_campaign_member_by_character_id(campaign_id, character_id)
  json_error(404, 'character not found') unless character

  json_error(403, 'forbidden') unless character[:owner] == username

  recipe = Storage.load_recipe(campaign_id, recipe_id)
  json_error(404, 'recipe not found') unless recipe

  result = Storage.craft_recipe(campaign_id, recipe_id, character_id)
  json_error(409, 'insufficient ingredients') unless result

  status 201
  JSON.dump(
    character_id: character_id,
    recipe_id: recipe_id,
    output_item: recipe[:output_item],
    output_quantity: recipe[:output_quantity]
  )
end
